### ISSUE TITLE
please have options -has_tag and -lacks_tag optionally support hierarchical tags

### ISSUE THREAD

**Issue by jonassmedegaard** [perlancar/perl-Regexp-Pattern#3]
Tags in Regexp::Pattern::License are hierarchical.
Would be nice if I could generate all patterns tagged type:license in the hierarchical sense - i.e. disregarding further subtags.

I propose this extension: If -has_tags or -lacks_tag ends with ":*" (colon asterix) then it is taken to mean "...or deeper".

E.g. like this:
```patch
--- Pattern.pm.orig
+++ Pattern.pm
@@ -97,10 +97,20 @@
             for my $n (sort keys %$var) {
                 my $tags = $var->{$n}{tags} || [];
                 if (defined $has_tag) {
-                    next unless grep { $_ eq $has_tag } @$tags;
+                    if ($has_tag =~ /:\*/) {
+                        my $has_tag_tag = substr $has_tag, 0, -2;
+                        next unless grep { $_ =~ /^\Q$has_tag_tag\E(?:\z|:)/ } @$tags;
+                    } else {
+                        next unless grep { $_ eq $has_tag } @$tags;
+                    }
                 }
                 if (defined $lacks_tag) {
-                    next if grep { $_ eq $lacks_tag } @$tags;
+                    if ($lacks_tag =~ /:\*/) {
+                        my $lacks_tag_tag = substr $lacks_tag, 0, -2;
+                        next if grep { $_ =~ /^\Q$lacks_tag_tag\E(?:\z|:)/ } @$tags;
+                    } else {
+                        next if grep { $_ eq $lacks_tag } @$tags;
+                    }
                 }
                 push @names, $n;
             }
```

**Comment by perlancar** [comment]
Would a regex match e.g. -has_tag_matching or -lacks_tag_matching (or -has_tag_pattern and -lacks_tag_pattern; other name suggestion welcome) be okay with you? That way, we don't have to assume any structure in the tag.

**Comment by jonassmedegaard** [comment]
Quoting perlancar (2020-02-07 16:04:42)
> Would a regex match e.g. -has_tag_matching or -lacks_tag_matching (or 
> -has_tag_pattern and -lacks_tag_pattern) other name suggestion 
> welcome) be okay with you? That way, we don't have to assume any 
> structure in the tag.

Yes, an explicit regex is fine.

-- 
 * Jonas Smedegaard - idealist & Internet-arkitekt
 * Tlf.: +45 40843136  Website: http://dr.jones.dk/

 [x] quote me freely  [ ] ask before reusing  [ ] keep private

**Comment by perlancar** [comment]
Done, released 0.2.12.

**Comment by jonassmedegaard** [comment]
Quoting perlancar (2020-02-07 16:52:35)
> Done, released 0.2.12.

Great. Thanks!

-- 
 * Jonas Smedegaard - idealist & Internet-arkitekt
 * Tlf.: +45 40843136  Website: http://dr.jones.dk/

 [x] quote me freely  [ ] ask before reusing  [ ] keep private
